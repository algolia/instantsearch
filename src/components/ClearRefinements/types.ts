import { ClearRefinementsRendererOptions } from '../../connectors/clear-refinements/connectClearRefinements';
import {
  ClearRefinementsCSSClasses,
  ClearRefinementsTemplates,
} from '../../widgets/clear-refinements/clear-refinements';

export type ClearRefinementsProps = {
  refine: ClearRefinementsRendererOptions['refine'];
  cssClasses: ClearRefinementsCSSClasses;
  hasRefinements: ClearRefinementsRendererOptions['hasRefinements'];
  templateProps: {
    [key: string]: any;
    templates: ClearRefinementsTemplates;
  };
};
