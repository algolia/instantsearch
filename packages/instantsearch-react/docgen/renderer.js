import marked from 'marked';
import {runMode} from 'codemirror/addon/runmode/runmode.node';
import 'codemirror/mode/jsx/jsx';
import escape from 'escape-html';

const highlight = source => {
  let output = '';
  // Even though highlight accepts a lang parameter, force to JSX for now.
  runMode(source, 'jsx', (text, style) => {
    text = escape(text);
    if (!style) {
      output += text;
      return;
    }
    const className = style.split(' ').map(s => `cm-${s}`).join(' ');
    output += `<span class="${className}">${text}</span>`;
  });
  return output;
};

const blockTheme = 'cm-s-default';
const spanTheme = 'cm-s-default';
const renderer = new marked.Renderer();
renderer.code = (code, lang) => {
  return `<pre class="CodeMirror ${blockTheme}"><code>${highlight(code, lang)}</code></pre>`;
};
renderer.codespan = code => {
  return `<code class="CodeMirror ${spanTheme}">${code}</code>`;
};

export default renderer;
