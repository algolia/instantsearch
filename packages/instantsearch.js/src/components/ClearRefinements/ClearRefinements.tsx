/** @jsx h */

import { cx } from '@algolia/ui-components-shared';
import { h } from 'preact';

import Template from '../Template/Template';

import type { ClearRefinementsRenderState } from '../../connectors/clear-refinements/connectClearRefinements';
import type { PreparedTemplateProps } from '../../lib/templating';
import type { ComponentCSSClasses } from '../../types';
import type {
  ClearRefinementsCSSClasses,
  ClearRefinementsTemplates,
} from '../../widgets/clear-refinements/clear-refinements';

export type ClearRefinementsComponentCSSClasses =
  ComponentCSSClasses<ClearRefinementsCSSClasses>;

export type ClearRefinementsComponentTemplates =
  Required<ClearRefinementsTemplates>;

export type ClearRefinementsProps = {
  refine: ClearRefinementsRenderState['refine'];
  cssClasses: ClearRefinementsComponentCSSClasses;
  hasRefinements: ClearRefinementsRenderState['hasRefinements'];
  templateProps: PreparedTemplateProps<ClearRefinementsComponentTemplates>;
};

const ClearRefinements = ({
  hasRefinements,
  refine,
  cssClasses,
  templateProps,
}: ClearRefinementsProps) => (
  <div className={cssClasses.root}>
    <Template
      {...templateProps}
      templateKey="resetLabel"
      rootTagName="button"
      rootProps={{
        className: cx(
          cssClasses.button,
          !hasRefinements && cssClasses.disabledButton
        ),
        onClick: refine,
        disabled: !hasRefinements,
      }}
      data={{ hasRefinements }}
    />
  </div>
);

export default ClearRefinements;
