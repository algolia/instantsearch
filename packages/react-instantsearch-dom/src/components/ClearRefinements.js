import classNames from 'classnames';
import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { translatable } from 'react-instantsearch-core';

import { createClassNames } from '../core/utils';

const cx = createClassNames('ClearRefinements');

class ClearRefinements extends Component {
  static propTypes = {
    items: PropTypes.arrayOf(PropTypes.object).isRequired,
    canRefine: PropTypes.bool.isRequired,
    refine: PropTypes.func.isRequired,
    translate: PropTypes.func.isRequired,
    className: PropTypes.string,
  };

  static defaultProps = {
    className: '',
  };

  render() {
    const { items, canRefine, refine, translate, className } = this.props;

    return (
      <div className={classNames(cx(''), className)}>
        <button
          className={cx('button', !canRefine && 'button--disabled')}
          onClick={() => refine(items)}
          disabled={!canRefine}
        >
          {translate('reset')}
        </button>
      </div>
    );
  }
}

export default translatable({
  reset: 'Clear all filters',
})(ClearRefinements);
