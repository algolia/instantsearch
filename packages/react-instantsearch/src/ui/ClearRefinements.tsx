import React from 'react';

import { cx } from './lib/cx';

export type ClearRefinementsTranslations = {
  /**
   * The label of the button
   */
  resetButtonText: string;
};

export type ClearRefinementsProps = React.ComponentProps<'div'> &
  Pick<React.ComponentProps<'button'>, 'disabled' | 'onClick'> & {
    translations: ClearRefinementsTranslations;
    classNames?: Partial<ClearRefinementsClassNames>;
  };

export type ClearRefinementsClassNames = {
  /**
   * Class names to apply to the root element
   */
  root: string;
  /**
   * Class names to apply to the button
   */
  button: string;
  /**
   * Class names to apply to the button when it's disabled
   */
  disabledButton: string;
};

export function ClearRefinements({
  classNames = {},
  disabled = false,
  onClick = () => {},
  translations,
  ...props
}: ClearRefinementsProps) {
  return (
    <div
      {...props}
      className={cx('ais-ClearRefinements', classNames.root, props.className)}
    >
      <button
        disabled={disabled}
        onClick={onClick}
        className={cx(
          'ais-ClearRefinements-button',
          classNames.button,
          disabled &&
            cx(
              'ais-ClearRefinements-button--disabled',
              classNames.disabledButton
            )
        )}
      >
        {translations.resetButtonText}
      </button>
    </div>
  );
}
