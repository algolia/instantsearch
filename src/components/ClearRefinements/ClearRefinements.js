import PropTypes from 'prop-types';
import React, { Component } from 'preact-compat';
import Template from '../Template';
import { isSpecialClick } from '../../lib/utils';

export default class RawClearRefinements extends Component {
  componentWillMount() {
    this.handleClick = this.handleClick.bind(this);
  }

  shouldComponentUpdate(nextProps) {
    return (
      this.props.url !== nextProps.url ||
      this.props.hasRefinements !== nextProps.hasRefinements
    );
  }

  handleClick(e) {
    if (isSpecialClick(e)) {
      // do not alter the default browser behavior
      // if one special key is down
      return;
    }
    e.preventDefault();
    this.props.refine();
  }

  render() {
    const { hasRefinements, cssClasses } = this.props;
    const data = { hasRefinements };

    const className =
      !hasRefinements && cssClasses.disabledButton
        ? `${cssClasses.button} ${cssClasses.disabledButton}`
        : cssClasses.button;

    return (
      <div className={cssClasses.root}>
        <Template
          rootTagName="button"
          rootProps={{
            className,
            href: this.props.url,
            onClick: this.handleClick,
            disabled: !hasRefinements,
          }}
          data={data}
          templateKey="button"
          {...this.props.templateProps}
        />
      </div>
    );
  }
}

RawClearRefinements.propTypes = {
  refine: PropTypes.func.isRequired,
  cssClasses: PropTypes.shape({
    root: PropTypes.string,
    button: PropTypes.string,
    disabledButton: PropTypes.string,
  }),
  hasRefinements: PropTypes.bool.isRequired,
  templateProps: PropTypes.object.isRequired,
  url: PropTypes.string.isRequired,
};
