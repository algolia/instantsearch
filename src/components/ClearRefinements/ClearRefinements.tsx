/** @jsx h */

import { h } from 'preact';
import cx from 'classnames';
import Template from '../Template/Template';
import type { ClearRefinementsRenderState } from '../../connectors/clear-refinements/connectClearRefinements';
import type {
  ClearRefinementsCSSClasses,
  ClearRefinementsTemplates,
} from '../../widgets/clear-refinements/clear-refinements';
import type { ComponentCSSClasses } from '../../types';
import type { PreparedTemplateProps } from '../../lib/utils';

export type ClearRefinementsComponentCSSClasses =
  ComponentCSSClasses<ClearRefinementsCSSClasses>;

export type ClearRefinementsComponentTemplates =
  Required<ClearRefinementsTemplates>;

export type ClearRefinementsProps = {
  refine: ClearRefinementsRenderState['refine'];
  cssClasses: ClearRefinementsComponentCSSClasses;
  hasRefinements: ClearRefinementsRenderState['hasRefinements'];
  templateProps: PreparedTemplateProps<ClearRefinementsComponentTemplates>;
};

const ClearRefinements = ({
  hasRefinements,
  refine,
  cssClasses,
  templateProps,
}: ClearRefinementsProps) => (
  <div className={cssClasses.root}>
    <Template
      {...templateProps}
      templateKey="resetLabel"
      rootTagName="button"
      rootProps={{
        className: cx(cssClasses.button, {
          [cssClasses.disabledButton]: !hasRefinements,
        }),
        onClick: refine,
        disabled: !hasRefinements,
      }}
      data={{ hasRefinements }}
    />
  </div>
);

export default ClearRefinements;
