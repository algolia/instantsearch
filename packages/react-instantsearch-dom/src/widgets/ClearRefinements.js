import React from 'react';
import { connectCurrentRefinements } from 'react-instantsearch-core';

import ClearRefinements from '../components/ClearRefinements';
import PanelCallbackHandler from '../components/PanelCallbackHandler';

/**
 * The ClearRefinements widget displays a button that lets the user clean every refinement applied
 * to the search.
 * @name ClearRefinements
 * @kind widget
 * @propType {function} [transformItems] - Function to modify the items being displayed, e.g. for filtering or sorting them. Takes an items as parameter and expects it back in return.
 * @propType {boolean} [clearsQuery=false] - Pass true to also clear the search query
 * @themeKey ais-ClearRefinements - the root div of the widget
 * @themeKey ais-ClearRefinements-button - the clickable button
 * @themeKey ais-ClearRefinements-button--disabled - the disabled clickable button
 * @translationKey reset - the clear all button value
 * @example
 * import React from 'react';
 * import algoliasearch from 'algoliasearch/lite';
 * import { InstantSearch, ClearRefinements, RefinementList } from 'react-instantsearch-dom';
 *
 * const searchClient = algoliasearch(
 *   'latency',
 *   '6be0576ff61c053d5f9a3225e2a90f76'
 * );
 * const App = () => (
 *   <InstantSearch
 *     searchClient={searchClient}
 *     indexName="instant_search"
 *   >
 *     <ClearRefinements />
 *     <RefinementList
 *       attribute="brand"
 *       defaultRefinement={['Apple']}
 *     />
 *   </InstantSearch>
 * );
 */

const ClearRefinementsWidget = (props) => (
  <PanelCallbackHandler {...props}>
    <ClearRefinements {...props} />
  </PanelCallbackHandler>
);

export default connectCurrentRefinements(ClearRefinementsWidget, {
  $$widgetType: 'ais.clearRefinements',
});
