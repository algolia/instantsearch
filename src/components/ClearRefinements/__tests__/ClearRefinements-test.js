import React from 'react';
import ClearRefinements from '../ClearRefinements';
import renderer from 'react-test-renderer';

describe('ClearRefinements', () => {
  const defaultProps = {
    refine: () => {},
    cssClasses: {
      button: 'custom-button',
      root: 'custom-root',
    },
    hasRefinements: true,
    templateProps: {
      templates: {
        button: '',
      },
    },
    url: '#all-cleared!',
  };

  it('should render <ClearRefinements />', () => {
    const tree = renderer
      .create(<ClearRefinements {...defaultProps} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a disabled <ClearRefinements />', () => {
    const props = {
      ...defaultProps,
      hasRefinements: false,
    };
    const tree = renderer.create(<ClearRefinements {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a disabled <ClearRefinements /> with a custom css class', () => {
    const props = {
      ...defaultProps,
      hasRefinements: false,
      cssClasses: {
        ...defaultProps.cssClasses,
        disabledButton: 'custom-disabled-button',
      },
    };
    const tree = renderer.create(<ClearRefinements {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should handle clicks (and special clicks)', () => {
    const props = {
      refine: jest.fn(),
    };
    const preventDefault = jest.fn();
    const component = new ClearRefinements(props);
    ['ctrlKey', 'shiftKey', 'altKey', 'metaKey'].forEach(e => {
      const event = { preventDefault };
      event[e] = true;
      component.handleClick(event);
      expect(props.refine).not.toHaveBeenCalled();
      expect(preventDefault).not.toHaveBeenCalled();
    });
    component.handleClick({ preventDefault });
    expect(props.refine).toHaveBeenCalledTimes(1);
    expect(preventDefault).toHaveBeenCalledTimes(1);
  });
});
