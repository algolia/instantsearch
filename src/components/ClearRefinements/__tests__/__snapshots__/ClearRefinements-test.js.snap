// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClearRefinements should render <ClearRefinements /> 1`] = `
<div
  className="custom-root"
>
  <button
    className="custom-button"
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
    disabled={false}
    href="#all-cleared!"
    onClick={[Function]}
  />
</div>
`;

exports[`ClearRefinements should render a disabled <ClearRefinements /> 1`] = `
<div
  className="custom-root"
>
  <button
    className="custom-button"
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
    disabled={true}
    href="#all-cleared!"
    onClick={[Function]}
  />
</div>
`;

exports[`ClearRefinements should render a disabled <ClearRefinements /> with a custom css class 1`] = `
<div
  className="custom-root"
>
  <button
    className="custom-button custom-disabled-button"
    dangerouslySetInnerHTML={
      Object {
        "__html": "",
      }
    }
    disabled={true}
    href="#all-cleared!"
    onClick={[Function]}
  />
</div>
`;
